controllers/helpers.go

```diff
@@ -113,3 +122,207 @@ func GetObjectsToRequestsByNamespaceAndClusterName(ctx context.Context, c client
+	AadClientID                  string `json:"aadClientId"`
+	AadClientSecret              string `json:"aadClientSecret"`
+	ResourceGroup                string `json:"resourceGroup"`
+	SecurityGroupName            string `json:"securityGroupName"`
+	Location                     string `json:"location"`
+	VMType                       string `json:"vmType"`
+	VnetName                     string `json:"vnetName"`
+	VnetResourceGroup            string `json:"vnetResourceGroup"`
+	SubnetName                   string `json:"subnetName"`
+	RouteTableName               string `json:"routeTableName"`
+	LoadBalancerSku              string `json:"loadBalancerSku"`
+	MaximumLoadBalancerRuleCount int    `json:"maximumLoadBalancerRuleCount"`
+	UseManagedIdentityExtension  bool   `json:"useManagedIdentityExtension"`
+	UseInstanceMetadata          bool   `json:"useInstanceMetadata"`
+	UserAssignedIdentityID       string `json:"userAssignedIdentityId"`
+}
+
+func reconcileAzureSecret(ctx context.Context, log logr.Logger, kubeclient client.Client, owner metav1.OwnerReference, new *corev1.Secret) error {
+	// Fetch previous secret, if it exists
+	key := types.NamespacedName{
+		Namespace: new.Namespace,
+		Name:      new.Name,
+	}
+	old := &corev1.Secret{}
+	err := kubeclient.Get(ctx, key, old)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return errors.Wrap(err, "failed to fetch existing azure json")
+	}
+
+	// Create if it wasn't found
+	if apierrors.IsNotFound(err) {
+		if err := kubeclient.Create(ctx, new); err != nil {
+			return errors.Wrap(err, "failed to create cluster azure json")
+		}
+		return nil
+	}
+
+	_, isGenerated := old.Labels[string(infrav1.ResourceLifecycleOwned)]
+
+	if !isGenerated {
+		log.Info("returning early from json reconcile, user provided secret already exists")
+		return nil
+	}
+
+	// Otherwise, check ownership and data fresheness. Update as necessary
+	hasOwner := false
+	for _, ownerRef := range old.OwnerReferences {
+		if referSameObject(ownerRef, new.OwnerReferences[0]) {
+			hasOwner = true
+			break
+		}
+	}
+
+	hasData := equality.Semantic.DeepEqual(old.Data, new.Data)
+	if hasData && hasOwner {
+		// no update required
+		log.Info("returning early from json reconcile, no update needed")
+		return nil
+	}
+
+	if !hasOwner {
+		old.OwnerReferences = append(old.OwnerReferences)
+	}
+
+	if !hasData {
+		old.Data = new.Data
+	}
+
+	log.Info("updating azure json")
+	if err := kubeclient.Update(ctx, old); err != nil {
+		return errors.Wrap(err, "failed to update cluster azure json when diff was required")
+	}
+
+	log.Info("done updating azure json")
+
+	return nil
+}
+
+// GetOwnerMachinePool returns the MachinePool object owning the current resource.
+func GetOwnerMachinePool(ctx context.Context, c client.Client, obj metav1.ObjectMeta) (*capiv1exp.MachinePool, error) {
+	for _, ref := range obj.OwnerReferences {
+		if ref.Kind == "MachinePool" && ref.APIVersion == capiv1exp.GroupVersion.String() {
```

Comment:
might want to drop the match on api version and just do Kind and Group

Reply:
+1, see #813